src/framework-ondata.js

```diff
@@ -1,18 +1,301 @@
+             var wrapper = this.nodeInstance.wrapper;
+
+             new Promise(function( resolve ) { 
+                 // Execute fRunUpdater which will also execute the updater
+                 resolve( wrapper.fRunUpdater( vni ) );
+
+             }).then( function() { 
+                 // fRunUpdater/Updater success path
+             
+                 // Returned OK, but updater could have set an error - check for that
+                 if ( stateChange( vni.errorState, lastErrorState ) ) {
+                    setOutputErrorFlag( vni );   
+                    lastOutputLm = undefined;  // got a new error so we will force sending output
+                 }
+
+                 handleOutput( outputPorts.output, lastOutputLm, vni.outputState() );
+                 handleError( vni, outputPorts.error,  lastErrorState.lm );
+
+             }, function( rejected ) { 
+                 // fRunUpdater/updater failed 
+                 if ( isInitState( vni.errorState ) ) { 
+                     // fRunUpdater/updater has not already set an error state - use the rejected info
+                     changeStateData( vni.errorState, rejected );
+                 }
+
+                 setOutputErrorFlag( vni, true ); 
+
+                 stateChange( vni.outputState, lastOutputLm );
+                 stateChange( vni.errorState, lastErrorState );
+
+                 handleOutput( outputPorts.output, lastOutputLm, vni.outputState() );
+                 handleError( vni, outputPorts.error, lastErrorState.lm );
+                  
+
+                 // If we haven't already processed the rejected error, do it now
+                 if ( rejected !== vni.errorState().data ) { 
+                     lastErrorState = _.clone( vni.errorState() );   
+                     changeStateData( vni.errorState, rejected );
+                     if ( changeStateData( vni.errorState, lastErrorState ) ) { 
+                         lastErrorState.lm = undefined; 
+                     } 
+                     handleError( vni, outputPorts.error, lastErrorState.lm );
+                 }
+
+             }).catch( function( e ) { 
+                 console.error( "framework-ondata unable to process fRunUpdater results!" );
+                 var err = ( e.stack ) ? e.stack : e;
+                 console.error( err );
+             }); 
+         } // have an fRunUpdater
+     } // not ready for fRunUpdater yet
 };
+
+/*******************************************************************************/
+/* Error & Output State handling                                               */
+/*******************************************************************************/                 
+
+/**
+ * change state data to the specified newData value
+ *
+ * @param stateFacade a getter/setterfunction for the current state
+ */ 
+function changeStateData( stateFacade, newData, newLm ) { 
+    var currentState = stateFacade();
+    currentState.data = newData;
+    if ( newLm ) { 
```

Comment:
newLm is used to determine if we need to create a new one or just use the one we already have.